cpp/src/arrow/array.cc

```diff
@@ -492,11 +492,52 @@ DictionaryArray::DictionaryArray(const std::shared_ptr<DataType>& type,
   SetData(data);
 }
 
+Status DictionaryArray::FromArrays(const std::shared_ptr<DataType>& type,
+                                   const std::shared_ptr<Array>& indices,
+                                   std::shared_ptr<Array>* out) {
+  if (indices->length() == 0) {
+    return Status::Invalid("Dictionary indices must have non-zero length");
+  }
+
+  DCHECK_EQ(type->id(), Type::DICTIONARY);
+  std::shared_ptr<DictionaryType> dict = std::static_pointer_cast<DictionaryType>(type);
```

Comment:
We usually do this as ` auto dict = static_cast<const DictionaryType&>(*type)`. That way you can avoid any additional overhead from constructing `std::shared_ptr`s.

Reply:
Also a bit more concise IMO

Reply:
Need `const auto&` otherwise the copy ctor will be invoked. I opened https://issues.apache.org/jira/browse/ARROW-2078 just now because I'm not sure we want types to be copyable

Reply:
Gah, I always forget about this particular detail of C++.